File: pages/home_page.py
from pages.base_page import BasePage


class HomePage(BasePage):
    SEARCH_INPUT = "input[type='text'], input[placeholder*='Search']"
    SEARCH_BUTTON = "button:has-text('Search')"

    def open(self, base_url: str) -> None:
        self.goto(base_url)

    def search_user(self, username: str) -> None:
        search_input = self.page.locator(self.SEARCH_INPUT).first
        search_input.fill(username)
        if self.page.locator(self.SEARCH_BUTTON).count() > 0:
            self.page.locator(self.SEARCH_BUTTON).first.click()
        else:
            search_input.press("Enter")
        # Wait for requested profile data to load before next step.
        self.page.locator(f"a[href*='github.com/{username}']").first.wait_for()

    def click_user_result(self, username: str) -> None:
        # This app auto-loads user details after search; avoid clicking external links.
        user_text = self.page.get_by_text(username, exact=False)
        if user_text.count() > 0:
            return

        repos_label = self.page.get_by_text("Repos", exact=False)
        if repos_label.count() > 0:
            return

        raise AssertionError(f"Could not find user result for '{username}'.")

File: pages/HomePage.ts
import type { Page } from '@playwright/test';
import { BasePage } from './BasePage.js';

export class HomePage extends BasePage {
  private static readonly SEARCH_INPUT =
    "input[type='text'], input[placeholder*='Search']";
  private static readonly SEARCH_BUTTON = "button:has-text('Search')";

  constructor(page: Page) {
    super(page);
  }

  async open(path = '/'): Promise<void> {
    await this.goto(path);
  }

  async searchUser(username: string): Promise<void> {
    const searchInput = this.page.locator(HomePage.SEARCH_INPUT).first();
    await searchInput.fill(username);
    const searchBtn = this.page.locator(HomePage.SEARCH_BUTTON);
    if ((await searchBtn.count()) > 0) {
      await searchBtn.first().click();
    } else {
      await searchInput.press('Enter');
    }
    await this.page.locator(`a[href*='github.com/${username}']`).first().waitFor();
  }

  async clickUserResult(username: string): Promise<void> {
    const userText = this.page.getByText(username, { exact: false });
    if ((await userText.count()) > 0) {
      return;
    }
    const repos = this.page.getByText('Repos', { exact: false });
    if ((await repos.count()) > 0) {
      return;
    }
    throw new Error(`Could not find user result for '${username}'.`);
  }
}
